rocAL/rocAL/source/hardware_video_decoder.cpp

```diff
@@ -0,0 +1,282 @@
+            status = Status::FAILED;
+            break;
+        }
+        if (ret == 0 && pkt.stream_index != _video_stream_idx) continue;
+        end_of_stream = (ret == AVERROR_EOF);
+        if (end_of_stream)
+        {
+            // null packet for bumping process
+            pkt.data = nullptr;
+            pkt.size = 0;
+        }
+
+        // submit the packet to the decoder
+        ret = avcodec_send_packet(_video_dec_ctx, &pkt);
+        if (ret < 0)
+        {
+            ERR("HardWareVideoDecoder::Decode Error while sending packet to the decoder\n");
+            status = Status::FAILED;
+            break;
+        }
+
+        // get all the available frames from the decoder
+        while (ret >= 0)
+        {
+            ret = avcodec_receive_frame(_video_dec_ctx, dec_frame);
+            if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) break;
+            if ((dec_frame->pts < select_frame_pts) || (ret < 0)) continue;
+            if (frame_count % stride == 0)
+            {
+                //retrieve data from GPU to CPU
+                if ((av_hwframe_transfer_data(sw_frame, dec_frame, 0)) < 0) {
+                    ERR("HardWareVideoDecoder::Decode avcodec_receive_frame() failed");
+                    return Status::FAILED;
+                }
+
+                dst_data[0] = out_buffer;
+                dst_linesize[0] = out_stride;
+                if (swsctx)
+                    sws_scale(swsctx, sw_frame->data, sw_frame->linesize, 0, sw_frame->height, dst_data, dst_linesize);
+                else
+                {
+                    // copy from frame to out_buffer
+                    memcpy(out_buffer, sw_frame->data[0], sw_frame->linesize[0] * out_height);
+                }
+                out_buffer = out_buffer + image_size;
+            }
+            ++frame_count;
+            av_frame_unref(sw_frame);
+            av_frame_unref(dec_frame);
+            if (frame_count == sequence_length * stride)  
+            {
+                sequence_filled = true;
+                break;
+            }
+        }
+        av_packet_unref(&pkt);
+        if (sequence_filled)  break;
+    } while (!end_of_stream);
+    avcodec_flush_buffers(_video_dec_ctx);
+    av_frame_free(&dec_frame);
+    av_frame_free(&sw_frame);
+    sws_freeContext(swsctx);
+    return status;
+}
+
+// Initialize will open a new decoder and initialize the context
+VideoDecoder::Status HardWareVideoDecoder::Initialize(const char *src_filename)
+{
+    VideoDecoder::Status status = Status::OK;
+    int ret;
+    AVDictionary *opts = NULL;
+
+    // open input file, and initialize the context required for decoding
+    _fmt_ctx = avformat_alloc_context();
+    _src_filename = src_filename;
+
+    // find if hardware decode is available
+    AVHWDeviceType hw_type = AV_HWDEVICE_TYPE_NONE;
+    hw_type = av_hwdevice_find_type_by_name("vaapi");
```

Comment:
pass the device_id here for initializing on the corresponding device

Reply:
We have access to 1 GPU HW machine. We are waiting for multi GPU machine to add this feature and check